From aed6392d8f73c739e4bf1e6ec1f90d47da616fe0 Mon Sep 17 00:00:00 2001
From: Jiri Pirko <jpirko@redhat.com>
Date: Tue, 8 Jul 2014 06:15:36 -0400
Subject: [netdrv] bonding: allow to add vlans on top of empty bond

Message-id: <1404800137-3945-1-git-send-email-jpirko@redhat.com>
Patchwork-id: 85293
O-Subject: [patch rhel7 net 1/2] bonding: allow to add vlans on top of empty bond
Bugzilla: 1048814
RH-Acked-by: Neil Horman <nhorman@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Veaceslav Falico <vfalico@redhat.com>

BZ1048814
https://brewweb.devel.redhat.com/taskinfo?taskID=7658069
Tested on my testing machine. Also tested with NM by jklimes.

upstream (net-next) commit 763e0ecd72fe90fdd73bb1aa1b72caf8381d2fff
Author: Jiri Pirko <jiri@resnulli.us>
Date:   Fri Jun 27 16:13:12 2014 +0200

    bonding: allow to add vlans on top of empty bond

    This limitation maybe had some reason in the past, but now there is not
    one -> removing this.

Signed-off-by: Jiri Pirko <jpirko@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/bonding/bond_main.c b/drivers/net/bonding/bond_main.c
index 2ac0e58..b1432d9 100644
--- a/drivers/net/bonding/bond_main.c
+++ b/drivers/net/bonding/bond_main.c
@@ -1327,12 +1327,6 @@ static netdev_features_t bond_fix_features(struct net_device *dev,
 
  read_lock(&bond->lock);
 
- if (!bond->first_slave) {
-  /* Disable adding VLANs to empty bond. But why? --mq */
-  features |= NETIF_F_VLAN_CHALLENGED;
-  goto out;
- }
-
  mask = features;
  features &= ~NETIF_F_ONE_FOR_ALL;
  features |= NETIF_F_ALL_FOR_ALL;
@@ -1344,7 +1338,6 @@ static netdev_features_t bond_fix_features(struct net_device *dev,
  }
  features = netdev_add_tso_features(features, mask);
 
-out:
  read_unlock(&bond->lock);
  return features;
 }
@@ -4380,13 +4373,6 @@ static void bond_setup(struct net_device *bond_dev)
  bond_dev->priv_flags |= IFF_BONDING;
  bond_dev->priv_flags &= ~(IFF_XMIT_DST_RELEASE | IFF_TX_SKB_SHARING);
 
- /* At first, we block adding VLANs. That's the only way to
-  * prevent problems that occur when adding VLANs over an
-  * empty bond. The block will be removed once non-challenged
-  * slaves are enslaved.
-  */
- bond_dev->features |= NETIF_F_VLAN_CHALLENGED;
-
  /* don't acquire bond device's netif_tx_lock when
   * transmitting */
  bond_dev->features |= NETIF_F_LLTX;
-- 
1.7.1